core/Inst.hpp

```diff
@@ -148,6 +163,13 @@ namespace olympia
         void setLast() { is_last_ = true; }
         bool getLast() const { return is_last_; }
 
+        // Set the STF iterator to rewind trace when flushing
+        void setSTFIterator(stf::STFInstReader::iterator it) { trace_it_ = it; }
+        stf::STFInstReader::iterator getSTFIterator() const { return std::get<stf::STFInstReader::iterator>(trace_it_); }
+
+        void setJSONIterator(uint64_t it) { trace_it_ = it; }
+        uint64_t getJSONIterator() const { return std::get<uint64_t>(trace_it_); }
```

Comment:
Can you give me a bit more guidance here? 

Where should I create the class to? Inside Inst?

If it was execution driven we would just use the PC to rewind?

Reply:
Sure. first thought: a local member variable that is a `tuple`.  Add a template method that returns the different type.

(Dunno if this will do the trick -- just brain storming)

```
// Defined in Generator header files
enum Generator { STF, JSON };
using JSONIterator = uint64_t;

// Defined in Inst.hpp
public:
    template<Generator gen>
    auto getInstIterator() const {
        return std::get<gen>(rewind_iterator_);
    }

private:
    std::tuple<stf::STFInstReader::iterator, JSONIterator> rewind_iterator_;
```
Usage:
```
   inst_ptr->getInstIterator<Generator::JSON>();
```

Reply:
> If it was execution driven we would just use the PC to rewind?

Not that simple.  Depending on how you implement exe mode, it might need more information than which PC do I need to go back to.  For example, if the functional components keep track of uncommitted changes (its own renames), then the functional engine might need more than the PC.  This is especially true if we're deep into a loop and the functional model has seen that PC more than once...

Reply:
OK makes sense.

Is there a reason why a tuple is preferred over a union like std::variant

Reply:
Nope, no reason.  That's actually better since we'll only support 1 input at a time...